### ISSUE TITLE
Separate RequestHandlers and Bitmap decoding

### ISSUE THREAD

**Issue by rharter** [square/picasso#1049]
This may be related to #916, but I'll let you decide.

Picasso's RequestHandler setup is very powerful and would be great in many other instances than just filling an ImageView.  

In my case, I want to copy the image at the URL returned from the Documents Provider into my file cache for future use. The Document Provider is a bit of a mess, since the result can be a ContentProvider Uri, or an HTTP Url, or a File Uri, or who knows what.  

I'd like to save the file at full resolution, but that can be problematic if I have to do that by loading the entire bitmap into memory, then write it out to a file. It would be much more convenient to get the InputStream from the URL and do what I want with it.

After some preliminary investigation, I think this can be accomplished by separating the `BitmapHunter`/`RequestHandler`'s bitmap decoding functionality into a separate step in the process.  So  the `RequestHandler` would return just an InputStream, and that would get passed to a `Decoder`, which could be a `BitmapDecoder`, or a `FileDecoder`.  

The `Target` could also be updated to be generic and handle whatever the result of the decoder is.

Here's what I see the API looking like

```java
Picasso.with(this)
  .load(imageUrl)
  .decode(new FileDecoder(outputFile)) // optional, defaults to new BitmapDecoder()
  // How would transformations fit in here?  Can't transform files, only Bitmaps.
  .into(new Target<File>() {
    ...
  });
```

Some questions I'd have here:

1. Is this too edge-casey for Picasso?  If so, I'll just pull the RequestHandler components out into another library.
1. How would transformations fit into this, since you can't transform files, only loaded bitmaps.  Perhaps if there are transformations, or the target is an ImageView, and the decoder is set to something other than a `BitmapDecoder`, that will be decoded in addition to a `BitmapDecoder`?  Sounds messy.
1. Would it make sense, instead of complicating the existing API for this functionality, to simply make the RequestHandler API public so it can be used independently of the other parts of Picasso?

**Comment by JohnWowUs** [comment]
Issue #506 (Save to file API) should handle this use case. It's still in the pipeline though.

**Comment by JakeWharton** [comment]
If you have a transformation we would have to decode the bitmap in order to run them.
```
load(url).save(file); // Bytes streamed to disk
load(url).resize(50, 50).save(file) // Bytes streamed to BitmapFactory, written to disk.
load(url).transform(circle()).save(file) // Bytes streamed to BitmapFactory, written to disk.
```
However, there's all kinds of problems when you start dealing with alternate actions.

What happens when I do the following?
```
load(url).save(file);
load(url).into(imageView);
```
Here's we want to stream bytes into the file as they're read, but we also want BitmapFactory to be streaming bytes off the network call and performing the decode.

This becomes _very_ challenging to do without Okio behind the scenes.

**Comment by rharter** [comment]
Yeah, I'll admit, I had the Okio architecture in mind when I thought this up.  The alternate action is why I was thinking of having a `Decoder` object in there.  So it would be something like this.

```
load(url).decode(new FileDecoder()).save(file);
load(url).decode(new BitmapDecoder()).into(imageView);
```

To take that a step further, my initial thought was not to have a `save` and `into`, but just into, with the imageView case basically being a convenience method that provides a default `Target<Bitmap>`.  So here, the target would be a generic type, and take whatever type the decoder outputs.

```
load(url).decode(new FileDecoder()).into(Target<File>() {...});
load(url).decode(new BitmapDecoder()).into(Target<Bitmap>() {...});
load(url).decode(new MetadataDecoder()).into(Target<Metadata>() {...});
```

In the case of `into(imageView)`, that convenience method would just call `into(Target<? extends Decoder> target)` passing it a `Target<Bitmap>` that simply sets the bitmap on the image view.

**Comment by JakeWharton** [comment]
Shortly after the debate that led to the v2 API, it became obvious that we really needed a low-level API and a high-level API—the latter of which was built on the former. This would provide the power and granularity of the underlying mechanisms (dispatcher, request handler, downloader) without the burden of the simple API for those which wanted to do more advanced things.

What you describe sounds very similar to what this low-level API could be. Maybe the distinction between the two doesn't have to be so binary, and could rather be a progressive-enhancement API. One thing I'm sure of is that we want to isolate users of the simple API from the more advanced methods. Perhaps just a different `load` method (like `buildRequest` or something) which returned a superset.

Anyways, I think it's worth exploring with the caveat that no one is really focusing on Picasso work. I'm more than happy to participate in driving the discussion around use cases, historical lessons, and the API, but I don't have much bandwidth for implementation.

I'll try to capture my thoughts on what the data pipeline is later tonight and post them here.

**Comment by JakeWharton** [comment]
Would be nice to get this architecture change in where decoding the bitmap is a target for bytes that's separate from obtaining those bytes.
